### PR TITLE
Pm: documentation: enhancements

### PR DESCRIPTION
**Pull request by ceolin**
- Add link to samples and tests
- Add missing device power states
- s/device idle/device runtime/g
- Add more information about pm policies
- Remove wrong information about device busy API

### CODE REVIEW

**Review by gmarull — COMMENTED** [review]
some nitpicks

File: doc/reference/power_management/index.rst [hunk 1]
```diff
@@ -432,7 +431,19 @@ Resume Device asynchronously API
    int pm_device_get(const struct device *dev);
 
 Marks the device as being used. This API will asynchronously
-bring the device to resume state. The API returns 0 on success.
+bring the device to resume state if it was suspended. If the device
+was already active, it just increment the device usage count.
```
Comment (erwango) [inline doc/reference/power_management/index.rst:456]:
```suggestion
was already active, it just increments the device usage count.
```

File: doc/reference/power_management/index.rst [hunk 2]
```diff
@@ -453,9 +464,19 @@ Suspend Device asynchronously API
 
    int pm_device_put(const struct device *dev);
 
-Marks the device as being released. This API asynchronously put
-the device to suspend state if not already in suspend state.
-The API returns 0 on success.
+Releases a device. This API asynchronously put the device to suspend
```
Comment (erwango) [inline doc/reference/power_management/index.rst:488]:
```suggestion
Releases a device. This API asynchronously puts the device to suspend
```

**Comment by ceolin** [comment]
Thanks @gmarull @erwango ! Would you mind take a second look, I've addressed the comments.

**Review by gmarull — APPROVED** [review]

**Comment by ceolin** [comment]
@nashif  can you take a look on it please.

File: doc/reference/power_management/index.rst [hunk 3]
```diff
@@ -138,6 +154,16 @@ power savings, and with a minimum residency value (in device tree, see
 :zephyr_file:`dts/bindings/power/state.yaml`) less than or equal to
 the scheduled system idle time duration.
 
+This policy also accounts the time necessary to become active
```
Comment (nashif) [inline doc/reference/power_management/index.rst:157]: accounts for the time

File: doc/reference/power_management/index.rst [hunk 4]
```diff
@@ -152,6 +178,9 @@ In this policy the application is free to decide which power state the
 system should go based on the remaining time for the next scheduled
 timeout.
 
+An example of application that defines its own policy can be found in
```
Comment (nashif) [inline doc/reference/power_management/index.rst:181]: An example of an application

File: doc/reference/power_management/index.rst [hunk 5]
```diff
@@ -379,44 +407,43 @@ Check Busy Status of All Devices API
...
+*******************************
 
-The Device Idle Power Management framework is a Active Power
+The Device Runtime Power Management framework is an Active Power
 Management mechanism which reduces the overall system Power consumtion
 by suspending the devices which are idle or not being used while the
 System is active or running.
 
-The framework uses device_set_power_state() API set the
+The framework uses :code:`pm_device_state_set()` API set the
```
Comment (nashif) [inline doc/reference/power_management/index.rst:418]: :c:func:\`pm_device_state_set\`

File: doc/reference/power_management/index.rst [hunk 6]
```diff
@@ -320,22 +359,9 @@ off, then such transactions would be left in an inconsistent state. This
...
-device power management is handled in a distributed manner rather than centrally in
-:code:`pm_system_suspend()`, the decision to enter deep sleep can be made based
-on whether all devices are already turned off.
-
-This feature can be also used to emulate a hardware feature found in some SOCs
-that causes the system to automatically enter deep sleep when all devices are idle.
-In such an usage, the busy status can be set by default and cleared as each
-device becomes idle. When :code:`pm_system_suspend()` is called, deep sleep can
-be entered if no device is found to be busy.
+When the :code:`pm_system_suspend()` is called, depending on the power state
```
Comment (nashif) [inline doc/reference/power_management/index.rst:362]: :c:func:`pm_device_suspend`

File: doc/reference/power_management/index.rst [hunk 7]
```diff
@@ -228,6 +257,16 @@ The four device power states:
...
 
+:code:`PM_DEVICE_STATE_SUSPENDING`
+
+   Device is currently transitioning from :code:`PM_DEVICE_STATE_ACTIVE` to
+   :code:`PM_DEVICE_STATE_SUSPEND`.
+
+:code:`PM_DEVICE_STATE_RESUMING`
+
+   Device is currently transitioning from :code:`PM_DEVICE_STATE_SUSPEND` to
+   :code:`PM_DEVICE_STATE_ACTIVE`.
```
Comment (nashif) [inline doc/reference/power_management/index.rst:268]: :c:macro:``PM_DEVICE_STATE_ACTIVE\`

File: doc/reference/power_management/index.rst [hunk 8]
```diff
@@ -447,9 +486,19 @@ Suspend Device asynchronously API
...
-The API returns 0 on success.
+Releases a device. This API asynchronously puts the device to suspend
+state if not already in suspend state if the usage count of this device
+reaches 0.
+
+Applications can monitor this operation waiting on the device
+condition variable :code:`dev->pm->condvar`. e.g:
+
+.. code-block:: c
+
```
Comment (nashif) [inline doc/reference/power_management/index.rst:497]: Applications should not be dealing with condition variables at all, this is all done on the device driver/device pm level.

**Review by nashif — CHANGES_REQUESTED** [review]

File: doc/reference/power_management/index.rst [hunk 9]
```diff
@@ -416,7 +443,19 @@ Resume Device asynchronously API
...
+bring the device to resume state if it was suspended. If the device
+was already active, it just increments the device usage count.
+The API returns 0 on success.
+
+Device drivers can monitor this operation waiting on the device
+condition variable :code:`dev->pm->condvar`. e.g:
+
+.. code-block:: c
+
+   k_mutex_init(&request_mutex);
```
Comment (nashif) [inline doc/reference/power_management/index.rst:455]: This all is an implementation detail, neither application nor driver should be dealing with this directly, this should be implemented in the runtime device power management and the power subsystem.

**Review by nashif — CHANGES_REQUESTED** [review]

File: doc/reference/power_management/index.rst (hunk 8 above)
Comment (nashif) [reply doc/reference/power_management/index.rst:487]: this was not addressed.

Comment (ceolin) [reply doc/reference/power_management/index.rst:487]: It was ... I replaced `applications` with`device drivers`.

```
Device drivers can monitor this operation waiting on the device
condition variable :code:`dev->pm->condvar`. e.g:
```

File: doc/reference/power_management/index.rst (hunk 9 above)
Comment (ceolin) [reply doc/reference/power_management/index.rst:455]: That is why I've introduced an API for this in https://github.com/zephyrproject-rtos/zephyr/pull/35315/commits/7d5f4d638f968db1911efad9637d068141668836

The problem is that this API is still not merged and what is documented here is for what we have in tree.
Note that prior these changes the way to do this was similar but waiting on k_poll_signal.

Now, if you are saying that the information about how a device should monitor an async request is not relevant then we have very different opinion. What I can do:

1) Keep it this way and update it in that other pr  (if this one is going to be merged before)
2) Change it and document future API waiting that pr to be merged
3) Remove pm_runtime documentation

Comment (nashif) [reply doc/reference/power_management/index.rst:455]: I suggest move this commit from this PR into the other PR and keep documentation close to the implementation.

**Comment by ceolin** [comment]
close due https://github.com/zephyrproject-rtos/zephyr/pull/35646

**Pull request closed without merging**